race-buffer/src/buffer.rs

```diff
@@ -0,0 +1,635 @@
+        tail
+    }
+
+    /// Create iterator over the entries currently present in the buffer without changing the tail
+    #[inline]
+    pub fn iter_peek<'b>(&'b mut self) -> RaceBufferIter<'a, 'b, E> {
+        let start_seqn = max(self.read_seqn, self.overwrite_seqn);
+        RaceBufferIter::new(self, start_seqn)
+    }
+
+    /// Get two slices which together represent the entries currently present in the buffer, where the second
+    /// slice comes directly after the first
+    pub fn get_linear_slices(&self) -> (&[E], &[E]) {
+        let overwrite_seqn_index = get_seqn_index(
+            self.capacity(),
+            self.overwrite_seqn,
+            self.use_base_2_indexing,
+        );
+        let write_seqn_index =
+            get_seqn_index(self.capacity(), self.write_seqn, self.use_base_2_indexing);
+        // Safe to assume entries in front of overwrite sequence number and behind write sequence number are initialized
+        if overwrite_seqn_index >= write_seqn_index
+            && (Into::<u64>::into(self.overwrite_seqn) != 0
+                || Into::<u64>::into(self.write_seqn) != 0)
+        {
+            unsafe {
+                (
+                    &*(&self.storage[overwrite_seqn_index..] as *const [MaybeUninit<E>]
+                        as *const [E]),
+                    &*(&self.storage[..write_seqn_index] as *const [MaybeUninit<E>]
+                        as *const [E]),
+                )
+            }
+        } else {
+            // Present entries do not cross end of storage, second slice has 0 length
+            unsafe {
+                (
+                    &*(&self.storage[overwrite_seqn_index..write_seqn_index]
+                        as *const [MaybeUninit<E>] as *const [E]),
+                    &*(&self.storage[write_seqn_index..write_seqn_index]
+                        as *const [MaybeUninit<E>] as *const [E]),
+                )
+            }
+        }
+    }
+
+    /// Get the number of items currently in the buffer which have not been read yet
+    pub fn len(&self) -> usize {
+        let start_seqn = self.read_seqn.max(self.overwrite_seqn);
+        let len: u64 = (self.write_seqn - start_seqn).into();
+        len as usize
+    }
+
+    /// Return true if there are no items to read
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Get capacity of buffer storage
+    #[inline]
+    pub fn capacity(&self) -> usize {
+        self.storage.len()
+    }
+
+    /// Get underlying storage slice
+    #[inline]
+    pub fn get_slice(&self) -> &[MaybeUninit<E>] {
+        self.storage
+    }
+
+    #[cfg(feature = "std")]
+    #[cfg(test)]
+    pub(crate) fn get_write_seqn(&self) -> SeqNum {
+        self.write_seqn
+    }
+
+    #[cfg(feature = "std")]
+    #[cfg(test)]
+    pub(crate) fn get_overwrite_seqn(&self) -> SeqNum {
+        self.overwrite_seqn
+    }
+}
+
+impl<E> Iterator for RaceBuffer<'_, E>
```

Comment:
In this case I would recommend making a dedicated iterator type that does the removal, created by a `drain(&mut self)` method on `RaceBuffer`.  This makes it extra clear that one is consuming the contents of the collection.  

For reference in the standard library: 
* https://doc.rust-lang.org/std/collections/binary_heap/struct.BinaryHeap.html#method.drain
* https://doc.rust-lang.org/std/collections/hash_map/struct.HashMap.html#method.drain